src/bos/operators/utils/clients/pcs.py

```diff
@@ -0,0 +1,231 @@
+    operation indicates what the desired operation should be, which is a string value containing
+    one or more of the supported transition names for the given hardware, e.g. 'on', 'off', or 'force-off'.
+
+    Once created, one of two responses are returned. A 2XX response results in a transition_start_output
+    object, or, an invalid request results in a 4XX and subsequent raised PCS exception.
+
+    Args:
+        xnames: an iterable of xnames
+        operation: A string/enum for what the nodes should transition to
+        task_deadline_minutes: How long should PCS operate on the nodes to bring them to complete the operation;
+          typecast to an integer value.
+        deputy_key: An optional string value that can be used to further handle instructing PCS to perform
+          state transitions on behalf of a known existing reservation.
+        session: An already existing session to use with PCS, if any
+
+    Returns:
+        A transition_start_output object, which is a record for the transition that was created. The most important
+        field of which is the 'transitionID' value, which allows subsequent follow-on to the created request.
+
+    Raises:
+        PowerControlException: Any non-nominal response from PCS, typically as a result of an unexpected payload
+          response, or a failure to create a transition record.
+    """
+    session = session or requests_retry_session()
+    try:
+        assert operation in set(['On', 'Off', 'Soft-Off', 'Soft-Restart', 'Hard-Restart', 'Init', 'Force-Off'])
+    except AssertionError:
+        raise PowerControlSyntaxException("Operation '%s' is not supported or implemented." %(operation))
+    params = {'location': [], 'operation': operation}
+    if task_deadline_minutes:
+        params['taskDeadlineMinutes'] = int(task_deadline_minutes)
+    for xname in xnames:
+        reserved_location = {'xname': xname}
+        if deputy_key:
+            reserved_location['deputyKey'] = deputy_key
+        params['location'].append(reserved_location)
+    response = session.post(TRANSITION_ENDPOINT, json=params)
+    try:
+        response.raise_for_status()
+        if not response.ok:
+            raise PowerControlException("Non-2XX response to power_status query: response %s: %s" %(response.status_c))
+    except requests.exceptions.HTTPError as err:
+        raise PowerControlException(err) from err
+    try:
+        return response.json()
+    except json.decoder.JSONDecodeError as jde:
+        raise PowerControlException(jde) from jde
+
+
+def power_on(nodes, session=None, task_deadline_minutes=1, **kwargs):
+    """
+    Sends a request to PCS for transitioning nodes in question to a powered on state.
+    Returns: A JSON parsed object response from PCS, which includes the created request ID.
+    """
+    session = session or requests_retry_session()
+    return _transition_create(xnames=nodes, operation='On', task_deadline_minutes=task_deadline_minutes,
```

Comment:
Does "task_deadline_minutes" actually have an impact, and if so should we tie it into the options for how long we're willing to wait for power-on/off before retries happen?

Reply:
It does, but there is a ton of nuance here because of how multiple operators can create requests, and actual power state is asynchronous to what has been done to act on the node. 1 is the lowest possible value that still makes sense (0 and -1 have special meaning).

Reply:
What I'm really asking is this: Our default for max_power_on_wait_time is two minutes.  Will having "task_deadline_minutes" set to 1 mean that PCS gives up after a minute and BOS will wait the extra minute for no reason?  Or does "task_deadling_minutes" not have any impact, possibly because we check the power state independently and don't care about the length of the "task".

Reply:
Its the latter, and the delicate balance is that ideally requests to PCS expire faster than we make new requests so that things do not back up. This is an integer value field, meaning we can't specify .5, or a 30 second timeout. This is the lowest possible setting for PCS request lengths that we can actually use. Regardless of this, PCS is coded to "accomplish" its task after it successfully communicates the transition to the node(s) in question, where success is defined as having communicated via redfish the request to change the power state, NOT a success about if the node actually ended up doing what was requested. This is a major consideration, and also the reason we don't particularly care if the requests go through, as we're only ever interested in the actual power state of the node. Having these requests be the smallest possible value allows PCS the opportunity to instruct the node to change power states, and the immediately clears the request so that future power actions can be taken on the node without holding up subsequent action.

Reply:
follow on for clarification: PCS offers zero garuntees that a transition request is honored, only that it did due diligence to pass the message on to the redfish interface in question. The problem with increasing this time to something larger means that a request could become queued between operators (power off force, then power on graceful, for example) because they actual power state can be verified asynchronously. The concern is that multiple force power off requests can be made by the power off operator, the first one eventually "takes", the power on operator says "ok, now power on" but PCS does not honor the request because of a race condition with the forceful power off operator retrying.

In testing, keeping this value as small as possible allows the requests pipes to remain as clean as possible, especially since we only ever care what actually happens to the node.

Reply:
Makes sense, thanks for the explanation.